cmov/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg"
+)]
+#![warn(missing_docs, rust_2018_idioms, unused_qualifications)]
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+use core::arch::asm;
+
+/// Move if zero.
+///
+/// Uses a `test` instruction to check if the given `condition` value is
+/// equal to zero, then calls `cmovz` (a.k.a. `cmove`) to conditionally move
+/// `src` to `dst` when `condition` is equal to zero.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[inline(always)]
+pub fn cmovz(condition: usize, src: usize, dst: &mut usize) {
+    unsafe {
+        asm! {
+            "test {0}, {0}",
+            "cmovz {1}, {2}",
+            in(reg) condition,
+            inlateout(reg) *dst,
+            in(reg) src
+        };
+    }
+}
+
+/// Move if not zero.
+///
+/// Uses a `test` instruction to check if the given `condition` value is not
+/// equal to zero, then calls `cmovnz` (a.k.a. `cmovne`) to conditionally move
+/// `src` to `dst` when `condition` is nonzero.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[inline(always)]
+pub fn cmovnz(condition: usize, src: usize, dst: &mut usize) {
+    unsafe {
+        asm! {
+            "test {0}, {0}",
+            "cmovnz {1}, {2}",
+            in(reg) condition,
+            inlateout(reg) *dst,
+            in(reg) src
+        };
+    }
+}
+
+/// Move if zero (portable fallback implementation).
+///
+/// This implementation is based on portable bitwise arithmetic but cannot
+/// guarantee that the resulting generated assembly is free of branch
+/// instructions.
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+#[inline(never)]
+pub fn cmovz(condition: usize, src: usize, dst: &mut usize) {
+    let zero_flag = 1 ^ is_non_zero(condition);
+    let src_mask = zero_flag.wrapping_mul(usize::MAX);
+    let dst_mask = zero_flag.wrapping_sub(1);
+    *dst = (*dst & dst_mask) | (src & src_mask);
+}
+
+/// Move if not zero (portable fallback implementation).
+///
+/// This implementation is based on portable bitwise arithmetic but cannot
+/// guarantee that the resulting generated assembly is free of branch
+/// instructions.
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+#[inline(never)]
+pub fn cmovnz(condition: usize, src: usize, dst: &mut usize) {
+    let nonzero_flag = is_non_zero(condition);
+    let src_mask = nonzero_flag.wrapping_mul(usize::MAX);
+    let dst_mask = nonzero_flag.wrapping_sub(1);
+    *dst = (*dst & dst_mask) | (src & src_mask);
+}
+
+/// Check if the given condition value is non-zero
+///
+/// # Returns
+/// - `condition` is zero: `0`
+/// - `condition` is non-zero: `1`
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+#[inline(always)]
+fn is_non_zero(condition: usize) -> usize {
+    const SHIFT_BITS: usize = core::mem::size_of::<usize>() - 1;
+    ((condition | (!condition).wrapping_add(1)) >> SHIFT_BITS) & 1
+}
```

Comment:
This can probably be improved, but seems to do the basic trick.

On x86, it emits the following assembly:

https://godbolt.org/z/eccaGzGsv

```asm
example::cmovz:
        mov     rax, rdi
        neg     rax
        or      rax, rdi
        shr     rax, 7
        mov     ecx, eax
        and     ecx, 1
        or      rax, -2
        neg     rcx
        and     rcx, qword ptr [rdx]
        add     rax, 1
        and     rax, rsi
        or      rax, rcx
        mov     qword ptr [rdx], rax
        ret

example::cmovnz:
        mov     eax, edi
        neg     eax
        or      eax, edi
        shr     eax, 7
        and     eax, 1
        mov     rcx, rax
        add     rax, -1
        and     rax, qword ptr [rdx]
        neg     rcx
        and     rcx, rsi
        or      rcx, rax
        mov     qword ptr [rdx], rcx
        ret
```

We could potentially lint that the generated assembly uses instructions from a specific allow list on the platforms cross tests run on.